lib/Tools/circt-bmc/LowerToBMC.cpp

```diff
@@ -0,0 +1,188 @@
+//===- LowerToBMC.cpp -----------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+
+#include "circt/Dialect/Seq/SeqTypes.h"
+#include "circt/Dialect/Verif/VerifOps.h"
+#include "circt/Support/Namespace.h"
+#include "circt/Tools/circt-bmc/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/FunctionCallUtils.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Location.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/APInt.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+
+namespace circt {
+#define GEN_PASS_DEF_LOWERTOBMC
+#include "circt/Tools/circt-bmc/Passes.h.inc"
+} // namespace circt
+
+//===----------------------------------------------------------------------===//
+// Convert Lower To BMC pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LowerToBMCPass : public circt::impl::LowerToBMCBase<LowerToBMCPass> {
+  using LowerToBMCBase::LowerToBMCBase;
+  void runOnOperation() override;
+};
+} // namespace
+
+static Value lookupOrCreateStringGlobal(OpBuilder &builder, ModuleOp moduleOp,
+                                        StringRef str) {
+  Location loc = moduleOp.getLoc();
+  auto global = moduleOp.lookupSymbol<LLVM::GlobalOp>(str);
+  if (!global) {
+    OpBuilder b = OpBuilder::atBlockEnd(moduleOp.getBody());
+    auto arrayTy = LLVM::LLVMArrayType::get(b.getI8Type(), str.size() + 1);
+    global = b.create<LLVM::GlobalOp>(
+        loc, arrayTy, /*isConstant=*/true, LLVM::linkage::Linkage::Private, str,
+        StringAttr::get(b.getContext(), Twine(str).concat(Twine('\00'))));
+  }
+
+  // FIXME: sanity check the fetched global: do all the attributes match what
+  // we expect?
+
+  return builder.create<LLVM::AddressOfOp>(loc, global);
+}
```

Comment:
Given that we always only call this exactly twice, we could also just create new unique globals each time and simplify this (i.e., the FIXME would also be dealt with). But also great as is for now.

Reply:
Yeah, that's a good point! I'll get that fixed in this PR since it's a quick one